CARDS-2594, CARDS-2595 and CARDS-2600: Unsubmitted cleanup fixes

sdumitriu opened the pull request:
CARDS-2594: UnsubmittedFormsCleanupTask may delete still valid forms
CARDS-2595: UnsubmittedFormsCleanupTask may fail if it has to delete too many forms
CARDS-2600: Delay the deletion of unsubmitted forms

sdumitriu committed:
CARDS-2594: UnsubmittedFormsCleanupTask may delete still valid forms

sdumitriu committed:
[misc] Speed up the UnsubmittedForms query by only looking at visit forms a few days before the cutoff date

sdumitriu committed:
CARDS-2595: UnsubmittedFormsCleanupTask may fail if it has to delete too many forms

sdumitriu committed:
CARDS-2600: Delay the deletion of unsubmitted forms

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/UnsubmittedFormsCleanupTask.java:
```diff
@@ -100,6 +100,11 @@ public void run()
                 }
+                // Leave two more months as a grace period in which we can identify problems
+                delay += 61;
+                ZonedDateTime upperLimit = DateUtils.atMidnight(ZonedDateTime.now()).minusDays(delay);
```
CARDS-2594: what was missing was the `atMidnight` call, which means that we were looking for visits with `time < now - 31 days`, which, given that these tasks take a really long time to run, might mean `time < 31 days ago at 13:00`, which included visits still valid for that day.

sdumitriu replied:
But given that we now add 61 more days, that shouldn't be relevant anymore, but it's more rigorous anyway and may become relevant again if we decide to remove the 61 days delay in the future.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/UnsubmittedFormsCleanupTask.java:
```diff
@@ -100,6 +100,11 @@ public void run()
                     delay = (int) clinicNode.getProperty("daysRelativeToEventWhileSurveyIsValid").getLong();
                 }
+                // Leave two more months as a grace period in which we can identify problems
+                delay += 61;
```
CARDS-2600: increase the delay by 61 days.

marta- replied:
Any reason why we're hardcoding this instead of adjusting the configurations?

sdumitriu replied:
Yes, I wanted to do something as quick as possible.

sdumitriu replied:
Let me see how quickly I can add a config for it.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/UnsubmittedFormsCleanupTask.java:
```diff
@@ -116,34 +121,35 @@ public void run()
                         LOGGER.warn("Failed to delete expired form {}: {}", form.getPath(), e.getMessage());
                     }
                 });
-                resolver.commit();
```
CARDS-2595: commit changes for each deleted form instead of all at once, since otherwise the session grows too big, too slow, and may OOM or timeout.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/UnsubmittedFormsCleanupTask.java:
```diff
@@ -100,6 +100,11 @@ public void run()
+                delay += 61;
+                ZonedDateTime upperLimit = DateUtils.atMidnight(ZonedDateTime.now()).minusDays(delay);
+                // Since this runs nightly, it's OK to look for the results just from a few days before
+                ZonedDateTime lowerLimit = upperLimit.minusDays(7);
```
Speedup, see comment above.

sdumitriu committed:
CARDS-2600: Delay the deletion of unsubmitted forms
Make the grace period configurable, along with the schedule expression and the option to not delete some questionnaires

marta- reviewed (approved):
Looks right, did not test

sdumitriu committed:
CARDS-2600: Delay the deletion of unsubmitted forms
Add configuration for YE
